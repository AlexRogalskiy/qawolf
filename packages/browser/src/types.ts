declare module "puppeteer/lib/USKeyboardLayout" {
  export interface KeyDefinition {
    code: string;
    key: string;
    keyCode?: number;
    location?: number;
    shiftKey?: string;
    shiftKeyCode?: number;
  }

  export interface KeyDefinitionsType {
    [key: string]: KeyDefinition;
  }

  const map: KeyDefinitionsType;
  export default map;
}
