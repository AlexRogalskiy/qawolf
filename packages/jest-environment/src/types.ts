import { Browser } from "@qawolf/browser";
import { Runner } from "@qawolf/runner";
import { AssertOptions, Step, StepValue, Workflow } from "@qawolf/types";
import { Page } from "puppeteer";

declare global {
  // declare the globals we expose in RunnerEnvironment
  const runner: Runner;

  function click(step: Step): Promise<void>;
  function hasText(text: string, options?: AssertOptions): Promise<boolean>;
  function input(step: Step, value?: StepValue): Promise<void>;
  function scrollElement(step: Step, value: StepValue): Promise<void>;

  const steps: Step[];
  const values: StepValue[];
  const workflow: Workflow;

  const browser: Browser;
  function currentPage(): Promise<Page>;
  function getPage(
    index?: number,
    waitForRequests?: boolean,
    timeoutMs?: number
  ): Promise<Page>;
}
