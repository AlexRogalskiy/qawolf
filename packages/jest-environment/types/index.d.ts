// setup types
declare var hello: string;
