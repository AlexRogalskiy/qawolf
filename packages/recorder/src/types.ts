export type RecordingSize = {
  height: number;
  width: number;
};

export type RecordingOffset = {
  x: number;
  y: number;
};
