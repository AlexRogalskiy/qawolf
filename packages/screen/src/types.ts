export type CaptureOffset = {
  x: number;
  y: number;
};

export type CaptureSize = {
  height: number;
  width: number;
};
