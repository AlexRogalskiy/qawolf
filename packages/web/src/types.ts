declare module "html-parse-stringify" {
  import { Doc } from "@qawolf/types";

  export interface IOptions {
    components: string[];
  }

  export function parse_tag(tag: string): Doc;

  export function parse(html: string, options?: IOptions): Array<any>;

  export function stringify(docs: Doc[]): string;
}
