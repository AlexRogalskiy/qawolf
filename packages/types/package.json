{
  "name": "@qawolf/types",
  "description": "qawolf types",
  "version": "0.4.3",
  "license": "BSD-3.0",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "files": [
    "lib",
    "src"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/qawolf/qawolf.git"
  },
  "bugs": {
    "url": "https://github.com/qawolf/qawolf/issues"
  },
  "publishConfig": {
    "access": "public"
  }
}
