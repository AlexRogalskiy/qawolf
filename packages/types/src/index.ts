export type Action = "click" | "type" | "scroll";

export type Callback<S = void, T = void> = (data?: S) => T;

export type ElementDescriptor = {
  ariaLabel?: string | null;
  classList?: string[] | null;
  dataValue?: string | null;
  href?: string | null;
  iconContent?: string[] | null;
  id?: string | null;
  inputType?: string | null;
  labels?: string[] | null;
  name?: string | null;
  parentText?: string[] | null;
  placeholder?: string | null;
  tagName?: string | null;
  innerText?: string | null;
  title?: string | null;
  xpath?: string | null;
};

export type EventName = "click" | "keyup" | "scroll";

export interface Event {
  name: EventName;
  isTrusted: boolean;
  pageId?: number;
  target: ElementDescriptor;
  time: number;
}

export interface KeyupEvent extends Event {
  name: "keyup";
  value?: string | null;
}

export type Locator = {
  action: Action;
  dataAttribute: string | null;
  target: ElementDescriptor;
  timeoutMs: number;
  value?: string | null;
};

export interface ScrollEvent extends Event {
  name: "scroll";
  value: ScrollValue;
}

export type ScrollValue = {
  x: number;
  y: number;
};

export type Size = "desktop" | "tablet" | "mobile";

export type Step = {
  action: Action;
  index: number;
  pageId?: number;
  target: ElementDescriptor;
  value?: StepValue;
};

export type StepValue = string | ScrollValue | null | undefined;

export type Workflow = {
  name: string;
  size: Size;
  steps: Step[];
  url: string;
};
