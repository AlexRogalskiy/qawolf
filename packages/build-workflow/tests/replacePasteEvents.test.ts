import { loadEvents } from "@qawolf/fixtures";
import { Event, PasteEvent, KeyEvent } from "@qawolf/types";
import {
  buildKeyEvents,
  findPasteKeyEvents,
  replacePasteEvents
} from "../src/replacePasteEvents";

let events: Event[];

beforeAll(async () => {
  events = await loadEvents("paste");
});

describe("buildKeyEvents", () => {
  it("builds key events for a paste event", () => {
    const keyEvents = buildKeyEvents({
      isTrusted: true,
      name: "paste",
      target: {},
      time: 0,
      value: "Hey!"
    });

    expect(
      keyEvents.map(k => `${k.name === "keydown" ? "↓" : "↑"}${k.value}`)
    ).toEqual([
      "↓Shift",
      "↓KeyH",
      "↑KeyH",
      "↑Shift",
      "↓KeyE",
      "↑KeyE",
      "↓KeyY",
      "↑KeyY",
      "↓Shift",
      "↓Digit1",
      "↑Digit1",
      "↑Shift"
    ]);
  });
});

describe("findPasteKeyEvents", () => {
  it("finds matching paste events", () => {
    const pasteIndex1 = events.findIndex(e => e.name === "paste");
    const pasteIndex2 = events.findIndex(
      (e, index) => index > pasteIndex1 && e.name === "paste"
    );

    const pasteEvents1 = findPasteKeyEvents(events, pasteIndex1);
    expect(pasteEvents1.map(e => e.time)).toEqual([
      1573155008974,
      1573155009100,
      1573155009219,
      1573155009218
    ]);

    const pasteEvents2 = findPasteKeyEvents(events, pasteIndex2);
    expect(pasteEvents2.map(e => e.time)).toEqual([
      1573155014683,
      1573155014815,
      1573155014894,
      1573155014894
    ]);
  });
});

describe("replacePasteEvents", () => {
  it("replaces CMD+V shortcuts with key events", () => {
    const replacedEvents = replacePasteEvents(events);
    expect(
      replacedEvents
        .filter(e => e.name === "keydown" || e.name === "keyup")
        .map(e => `${e.name === "keydown" ? "↓" : "↑"}${(e as KeyEvent).value}`)
    ).toMatchSnapshot();
  });
});
