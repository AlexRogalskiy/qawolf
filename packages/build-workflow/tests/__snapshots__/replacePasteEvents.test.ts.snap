// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replacePasteEvents replaces CMD+V shortcuts with key events 1`] = `
Array [
  "↓Shift",
  "↓KeyS",
  "↑KeyS",
  "↑Shift",
  "↓KeyU",
  "↑KeyU",
  "↓KeyP",
  "↑KeyP",
  "↓KeyE",
  "↑KeyE",
  "↓KeyR",
  "↑KeyR",
  "↓Shift",
  "↓KeyS",
  "↑KeyS",
  "↑Shift",
  "↓KeyE",
  "↑KeyE",
  "↓KeyC",
  "↑KeyC",
  "↓KeyR",
  "↑KeyR",
  "↓KeyE",
  "↑KeyE",
  "↓KeyT",
  "↑KeyT",
  "↓Shift",
  "↓KeyP",
  "↑KeyP",
  "↑Shift",
  "↓KeyA",
  "↑KeyA",
  "↓KeyS",
  "↑KeyS",
  "↓KeyS",
  "↑KeyS",
  "↓KeyW",
  "↑KeyW",
  "↓KeyO",
  "↑KeyO",
  "↓KeyR",
  "↑KeyR",
  "↓KeyD",
  "↑KeyD",
  "↓Shift",
  "↓Digit1",
  "↑Digit1",
  "↑Shift",
  "↓MetaLeft",
  "↓KeyA",
  "↑MetaLeft",
  "↓Backspace",
  "↑Backspace",
  "↓KeyT",
  "↑KeyT",
  "↓KeyO",
  "↑KeyO",
  "↓KeyM",
  "↑KeyM",
  "↓KeyS",
  "↑KeyS",
  "↓KeyM",
  "↑KeyM",
  "↓KeyI",
  "↑KeyI",
  "↓KeyT",
  "↑KeyT",
  "↓KeyH",
  "↑KeyH",
  "↓Shift",
  "↓KeyS",
  "↑KeyS",
  "↑Shift",
  "↓KeyU",
  "↑KeyU",
  "↓KeyP",
  "↑KeyP",
  "↓KeyE",
  "↑KeyE",
  "↓KeyR",
  "↑KeyR",
  "↓Shift",
  "↓KeyS",
  "↑KeyS",
  "↑Shift",
  "↓KeyE",
  "↑KeyE",
  "↓KeyC",
  "↑KeyC",
  "↓KeyR",
  "↑KeyR",
  "↓KeyE",
  "↑KeyE",
  "↓KeyT",
  "↑KeyT",
  "↓Shift",
  "↓KeyP",
  "↑KeyP",
  "↑Shift",
  "↓KeyA",
  "↑KeyA",
  "↓KeyS",
  "↑KeyS",
  "↓KeyS",
  "↑KeyS",
  "↓KeyW",
  "↑KeyW",
  "↓KeyO",
  "↑KeyO",
  "↓KeyR",
  "↑KeyR",
  "↓KeyD",
  "↑KeyD",
  "↓Shift",
  "↓Digit1",
  "↑Digit1",
  "↑Shift",
]
`;
